### PR TITLE
Hack week: Update target SKD to 33

### PR DESCRIPTION
**Pull request by 0nko**
This PR updates the target SDK to version 33 and implements the new granular media access permissions (images, video, audio), described in the [documentation](https://developer.android.com/about/versions/13/behavior-changes-13).

On devices running older Android versions (< API 33), the original single `Storage` permission is requested, which gives access to all media. On versions 33 and up, new granular permissions are requested (images and video are requested in a single permission request, audio in another one).

| Old version | New version |
| --- | --- |
| image | image |

**To test:**

To test this PR, please prepare 2 devices/emulators with Android 33 and <33. Also make sure the devices have at least one image to display something (you can take a picture in the Camera app).

_API 33:_
1. Start with a clean installation of the sample app
2. Tap on the Local Device Picker button
3. Tap on the Allow button
4. Notice a permission request for `music and audio` appears
5. Grant the access
6. Notice another permission request appears, now for `photos and videos`
7. Grant the access
8. Notice the images stored on a device are listed
9. If you clear the app storage and try this again but deny the permissions (twice), you'll see the screen show on the screenshot above

_API <33:_
1. Repeat the same steps as for API 33
2. When you tap on the Allow button, you'll get a single permission request mentioning `photos, media and files`
3. If you deny the permission, you'll see the screen above

### CODE REVIEW

File: mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/model/MediaType.kt [hunk 1]
```diff
@@ -8,5 +9,8 @@ enum class MediaType {
 }
 
 enum class MediaTypes(val allowedTypes: Set<MediaType>) {
-    IMAGES(setOf(IMAGE)), VIDEOS(setOf(VIDEO)), IMAGES_AND_VIDEOS(setOf(IMAGE, VIDEO))
+    IMAGES(setOf(IMAGE)),
+    VIDEOS(setOf(VIDEO)),
```
Comment (irfano) [inline mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/model/MediaType.kt:13]: Is a single AUDIO type not necessary for the project?

File: mediapicker/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -3,6 +3,9 @@
     package="org.wordpress.android.mediapicker">
     <uses-permission android:name="android.permission.CAMERA" />
     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
+    <uses-permission android:name="android.permission.READ_MEDIA_IMAGES" />
+    <uses-permission android:name="android.permission.READ_MEDIA_VIDEO" />
+    <uses-permission android:name="android.permission.READ_MEDIA_AUDIO" />
     <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" android:maxSdkVersion="28" />
```
Comment (irfano) [inline mediapicker/src/main/AndroidManifest.xml:9]: > <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" android:maxSdkVersion="28" />

`android:maxSdkVersion="28"` seems wrong. It [should](https://developer.android.com/reference/android/Manifest.permission#WRITE_EXTERNAL_STORAGE) be `29`. It's not related to API 33 migration, but I think we can fix it here.

File: mediapicker/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -3,6 +3,9 @@
     package="org.wordpress.android.mediapicker">
     <uses-permission android:name="android.permission.CAMERA" />
     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
```
Comment (irfano) [inline mediapicker/src/main/AndroidManifest.xml]: We can add `android:maxSdkVersion="32"` on [READ_EXTERNAL_STORAGE](https://developer.android.com/reference/android/Manifest.permission#READ_EXTERNAL_STORAGE) permission.

File: sampleapp/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -2,6 +2,7 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="org.wordpress.android.sampleapp">
 
+    <uses-permission android:name="android.permission.READ_MEDIA_IMAGES" />
```
Comment (irfano) [inline sampleapp/src/main/AndroidManifest.xml]: Should we also add `READ_MEDIA_IMAGES` here? I think we can remove it since it's already added to the library's `AndroidManifest`.

File: mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt [hunk 5]
```diff
@@ -577,18 +597,46 @@ internal class MediaPickerViewModel @Inject constructor(
         }
     }
 
+    fun checkMediaPermissions(permissions: List<PermissionsRequested>, isAlwaysDenied: Boolean) {
+        if (!mediaPickerSetup.isImagesPermissionRequired &&
+            !mediaPickerSetup.isVideoPermissionRequired &&
+            !mediaPickerSetup.isAudioPermissionRequired
+        ) {
```
Comment (irfano) [inline mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt]: Instead of checking the requirements of all media permissions, should we check only for `permissions` that come as the function parameter?

I guess this screen supports only requesting all media permissions. We can't request just "Audio". Is having the ability to request specific media permission not necessary for the project?

File: mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt [hunk 6]
```diff
@@ -600,38 +648,125 @@ internal class MediaPickerViewModel @Inject constructor(
...
+                )
+            }
+        } else {
+            buildOldSoftAskView(softAskRequest)
+        }
+    }
+
+    // Requests READ_EXTERNAL_STORAGE on devices running SDK < 33
+    @Suppress("NestedBlockDepth")
+    private fun buildOldSoftAskView(softAskRequest: SoftAskRequest?): SoftAskViewUiModel {
```
Comment (irfano) [inline mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt]: "buildOldSoftAskView" and "buildMediaSoftAskRequest" share many common lines of code. What do you think about merging them? Have you considered it?"

**Review by irfano — COMMENTED** [review]
I left some comments and questions, and I haven't approved the PR yet in case you want to make changes.

File: mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/model/MediaType.kt (hunk 1 above)
Comment (0nko) [reply mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/model/MediaType.kt:13]: It's not used anywhere, AFAIK. However, `MediaTypes` enum is just a helper for the common options, you can specify whatever you want because `MediaPickerSetup` takes `Set<MediaType>`.

File: mediapicker/src/main/AndroidManifest.xml (hunk 2 above)
Comment (0nko) [reply mediapicker/src/main/AndroidManifest.xml:9]: Nice catch.

Comment (0nko) [reply mediapicker/src/main/AndroidManifest.xml:9]: Actually, `28` is correct here. In API 29 you could have used it to `requestLegacyExternalStorage` but that's not our case. There is a nice explanation [here](https://stackoverflow.com/questions/64221188/write-external-storage-when-targeting-android-10) (what a mess, lol).

File: mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt (hunk 5 above)
Comment (0nko) [reply mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt]: Yes, that makes sense. This was the result of copying and refactoring the original implementation.

Comment (0nko) [reply mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt]: > Is having the ability to request specific media permission not necessary for the project?

You can do that, you just have to create the `MediaPickerSetup` instance yourself instead of using the `DeviceMediaPickerSetup`, for example. See my other [comment](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/pull/63#discussion_r1153450761).

File: mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt (hunk 6 above)
Comment (0nko) [reply mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt]: I took a closer look at this and I've done a bit more refactoring to simplify things. I just need to fix some edge cases and test things properly so please hang tight :).

File: mediapicker/src/main/AndroidManifest.xml (hunk 2 above)
Comment (irfano) [reply mediapicker/src/main/AndroidManifest.xml:9]: Thank you for the explanation. I was not aware of it. 👍🏻

**Comment by 0nko** [comment]
@irfano Thanks for your patience. I went back and streamlined the way permissions are being requested and ended up refactoring and removing some redundant code. Ready for another round.

File: mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt [hunk 7]
```diff
@@ -546,49 +561,85 @@ internal class MediaPickerViewModel @Inject constructor(
...
+                }
+                PermissionsRequested.MUSIC -> {
+                    mediaPickerSetup.isAudioPermissionRequired &&
+                            permissionsHandler.hasAudioPermission()
+                }
+                else -> throw UnsupportedOperationException("Unsupported permission: $it")
+            }
+        }
+
+        if (haveAllPermissions) {
```
Comment (irfano) [inline mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt]: We can use `hasAllRequiredPermissions()` here.

File: mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/api/MediaPickerSetup.kt [hunk 8]
```diff
@@ -4,19 +4,20 @@ import android.content.Intent
...
 import org.wordpress.android.mediapicker.model.MediaType
 import org.wordpress.android.mediapicker.model.MediaType.AUDIO
 import org.wordpress.android.mediapicker.model.MediaType.IMAGE
 import org.wordpress.android.mediapicker.model.MediaType.VIDEO
 
 data class MediaPickerSetup(
     val primaryDataSource: DataSource,
-    val availableDataSources: Set<DataSource>,
     val isMultiSelectEnabled: Boolean,
-    val needsAccessToStorage: Boolean,
```
Comment (irfano) [inline mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/api/MediaPickerSetup.kt]: `needsAccessToStorage` looks like actually necessary here. Not all image pickers require image permission, such as the gif media picker. I discovered an issue where the gif media picker screen shows the permission message unnecessarily.

<image src="https://user-images.githubusercontent.com/2471769/229884121-b17e37be-5181-4367-a74d-fb01e24fab18.png" width=200>

File: mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/api/MediaPickerSetup.kt [hunk 9]
```diff
@@ -27,8 +28,11 @@ data class MediaPickerSetup(
         HIDDEN, VISIBLE_TOGGLED, VISIBLE_UNTOGGLED
     }
 
-    val isStoragePermissionRequired = Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU &&
-        needsAccessToStorage
+    val isReadStoragePermissionRequired = when {
+        primaryDataSource == DEVICE && Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU -> true
+        primaryDataSource == CAMERA && Build.VERSION.SDK_INT < Build.VERSION_CODES.Q -> true
```
Comment (irfano) [inline mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/api/MediaPickerSetup.kt]: Can you help me to understand why we check `Build.VERSION.SDK_INT < Build.VERSION_CODES.Q` for CAMERA here? 🤔
I used Build.VERSION_CODES.R in [WPAndroid's PR](https://github.com/wordpress-mobile/WordPress-Android/pull/18183/files#diff-5aae92a41306a6b5a8aa97f3f41405b5c2ec6513b4bf39f8bcd6f65be17ee3ceR681-R687).

**Review by irfano — CHANGES_REQUESTED** [review]
Thank you for the new commits in response to my comment. I have reviewed all of them, and they have resolved my concerns. 
I have added some further comments and questions for your attention.

File: mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt (hunk 7 above)
Comment (0nko) [reply mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt]: Good call 👍

Comment (0nko) [reply mediapicker/src/main/java/org/wordpress/android/mediapicker/viewmodel/MediaPickerViewModel.kt]: Fixed [here](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/pull/63/commits/06ab4d035e2df3b226c44ab944cd5635337d163a).

File: mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/api/MediaPickerSetup.kt (hunk 8 above)
Comment (0nko) [reply mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/api/MediaPickerSetup.kt]: You're right, the storage access permissions is relevant only to the local device picker. I added a condition to the requirement check for [all media types](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/pull/63/files#diff-9c3d74783b3a9de70c8f6e388f81005f19327ea0a3cfdfc9a234dcae1196d997R35) for that, so we still don't need to provide `needsAccessToStorage` explicitly.

File: mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/api/MediaPickerSetup.kt (hunk 9 above)
Comment (0nko) [reply mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/api/MediaPickerSetup.kt]: The way this has been implemented was a bit confusing -- we checked for `READ_EXTERNAL_STORAGE` for camera but then requested `WRITE_EXTERNAL_STORAGE`. I've changed it now.

> Can you help me to understand why we check Build.VERSION.SDK_INT < Build.VERSION_CODES.Q for CAMERA here? 🤔

As for the API level required for the `WRITE_EXTERNAL_STORAGE`, it's outlined in the [Stack Overflow](https://stackoverflow.com/questions/64221188/write-external-storage-when-targeting-android-10) answer I mentioned earlier: You only need this permission in API 29 if you're using the `requestLegacyExternalStorage` flag in Android 10 (API 29). We don't do that, so we only require it up to API 28 (hence Build.VERSION.SDK_INT < Build.VERSION_CODES.Q condition). The check is now a bit different [here](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/pull/63/commits/333deb84caa239b62738ccaae8a85c4b1cdc7464#).

**Comment by 0nko** [comment]
Hey @irfano, I've addressed your comments and the PR's ready for a review.

File: mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/api/MediaPickerSetup.kt (hunk 9 above)
Comment (irfano) [reply mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/api/MediaPickerSetup.kt]: Thank you for explaining it.

**Review by irfano — APPROVED** [review]
Thank you for all of your work and your patience with my review. 😁 
I tested different `MediaPickerSetup`s (only `IMAGES` and `EVERYTHING`), and both worked great. LGTM! 👍🏻
